dlna_dms fixes from code review

chishm opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Code quality improvements motivated by @MartinHjelmare's [code review](https://github.com/home-assistant/core/pull/66437#pullrequestreview-895986032) of the initial `dlna_dms` pull request.

Also removes unnecessary calls to `BrowseMediaSource.calculate_children_class` as [noted on a different PR](https://github.com/home-assistant/core/pull/67409#discussion_r816971349).

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [x] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

chishm committed:
Fix issues found in code review

chishm committed:
Remove BrowseMediaSource.calculate_children_class call

It's called at the site of need, in BrowseMedia.as_dict

MartinHjelmare commented on tests/components/dlna_dms/conftest.py:
```diff
@@ -129,3 +128,41 @@ def ssdp_scanner_mock() -> Iterable[Mock]:
+    assert await hass.config_entries.async_setup(config_entry_mock.entry_id)
+    await hass.async_block_till_done()
+
+    domain_data = cast(DlnaDmsData, hass.data[DOMAIN])
```
We shouldn't need to access hass.data in the tests. We don't want to interact with any integration details.

We can assert core state, library mocks or eg return values from the media_source integration top level api.

https://developers.home-assistant.io/docs/development_testing#writing-tests-for-integrations

chishm replied:
I think I get where you're coming from. We want to black-box test the integrations as much as possible.

However, there's a few parts in `dlna_dms` where the test is done better by accessing the internal data directly. For example, checking invariants related to `DmsDeviceSource.source_id` and `DlnaDmsData.sources` is a good thing to do during the name change test, but can't be done without accessing the domain data. Most integrations wouldn't need to worry about that because they have entities handled by entity platforms. [Media sources are not entities](https://github.com/home-assistant/architecture/discussions/675#discussioncomment-1744709), so some things are done manually in `dlna_dms`.

I've changed most tests where it can be done via the `media_source` interface, but left some using the domain data directly for reasons outlined above.

MartinHjelmare replied:
I don't agree. How is the name change expressed in core state, library interface or media source interface or other interfaces that matter.

chishm replied:
I think that you should add motivations and explanations to the "Writing tests for integrations" section. Currently the only reason is "Following this pattern will make the tests more robust for changes in the integration", but it doesn't say how following the commandments will accomplish that. How is using `hass.data` going to break things? Why not unit-test the integration components, as well as black-box test the integration as a whole?

To rewrite `test_dlna_dms_data.py::test_update_source_id` so it doesn't touch `hass.data` I would have to:
1. Add a 2nd config entry to change the behaviour of `DmsMediaSource.async_browse_media` so that I could get the source ID(s) out via the `media_player.async_browse_media` call.
2. Call `media_player.async_browse_media` on the old and new source IDs to see that the dict keys have shifted
3. Catch the resulting `BrowseError`s and match on their strings to see if the error is because the source ID doesn't exist or because it does exist but the device mock is "disconnected".

This makes it more fragile, as now the test relies on the exact top-level browsing behaviour (1 above) and the exact error message strings (3). And after all that I still wouldn't know if there's any bad keys added to `DlnaDmsData.sources` by the name-change algorithm, because the test can't peek into the dict.

The only interfaces available are `media_source.async_browse_media` and `media_player.async_resolve_media`. There's no core state. The library doesn't know nor care what we call the sources.

MartinHjelmare replied:
I looked a bit at the source id code. I think the main problem is that we use the config entry title as a source identifier instead of using a stable id. Why do we not use the config entry entry_id as eg the motion_eye integration does?

chishm replied:
For `dlna_dms`, the [`source_id`s](https://www.home-assistant.io/integrations/dlna_dms/#media-source-uris) are used like an `entity_id` would be in most other service calls: to distinguish which entity (DMS server) is being used. They are user-visible and can be used in `media_player.play_media` service calls. I made sure that media items could be specified in the service call without having to select the item via the HA frontend, and that all parts of the `media-source://dlna_dms/` URI could be hand-written without knowledge of UUIDs.

I had a look at the motionEye integration. Its media source capabilities seem to be mostly about providing a browsable view of captured events within the frontend, or taking events and piping stored file URLs directly to a media player.

Compare `"media-source://dlna_dms/my_server/videos/favourites/Epic Sax Guy 10 Hours.mp4` to `media-source://motioneye/74565ad414754616000674c87bdc876c#662aa1c77657dbc4af836abcdf80000a#movies#/2021-04-10/21-27-53.mp4` for someone trying to generate the URL in a `media_player.play_media` service call.

balloob replied:
We can't use user generated content (config entry title) in any identifier that is meant to be stable. Now renaming config entry has unintended side effects and causes frustration.

If you want a stable identifier, you could take the server name, slugify it and store it in config entry data and then use that as a stable identifier (which cannot be changed).

chishm replied:
Ignoring the fact that [`entity_id` can be changed](https://www.home-assistant.io/docs/configuration/customizing-devices/#changing-the-entity-id), I've solidified the `source_id`s in the config_entry and written code to add them to existing entries.

chishm replied:
I've also reduced the use of `hass.data` to checking for resource clean up, and for one test where it's too hard to get the timing right via SSDP and asyncio.

chishm committed:
Black-box test dlna_dms where it makes sense

chishm committed:
Do not allow changing of source_id

It is fixed in the config entry now, and cannot be changed from the
pre-generated value (based on the original device name).

chishm committed:
Drop unused UpnpEventHandler

The async_upnp_client library has been updated to not require setting an
event_handler for the DmsDevice class, so don't create one.

chishm committed:
The DmsDeviceSource.location should never be empty

chishm committed:
Minimal interaction with dlna_dms internals while testing

The internal data for `dlna_dms` is only used for checking object cleanup
and for removing timing-dependency from one test.

chishm committed:
Cleanup tests

chishm committed:
Fix import problems found by flake8 & isort

MartinHjelmare commented on homeassistant/components/dlna_dms/__init__.py:
```diff
@@ -26,3 +27,21 @@ async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     return await get_domain_data(hass).async_unload_entry(entry)
+
+
+async def async_migrate_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
```
If we can do a soft migration without increasing the config entry version we prefer that, as it allows roll backs. We could instead check if the source id key is missing in the config entry data, and update it if needed, during config entry setup.

chishm replied:
Will do.

MartinHjelmare commented on homeassistant/components/dlna_dms/dms.py:
```diff
@@ -71,26 +65,19 @@ def __init__(
     async def async_setup_entry(self, config_entry: ConfigEntry) -> bool:
         """Create a DMS device connection from a config entry."""
         assert config_entry.unique_id
         async with self.lock:
```
Why do we need a lock here?

chishm replied:
Good question. It was a hold-over from several revisions ago and can be removed.

chishm committed:
Soft-migrate config entry to allow roll-backs

chishm committed:
Remove unneeded lock from DlnaDmsData

chishm committed:
Merge branch 'dev' into dlna_dms-review_fixes

chishm commented:
Thanks for the review, @MartinHjelmare 

I've merged in the dev branch and resolved the conflicts. I'm pretty sure I've addressed all of the review comments and made the requested changes, so this should be ready to merge.

MartinHjelmare commented on homeassistant/components/dlna_dms/dms.py:
```diff
@@ -71,22 +69,19 @@ def __init__(
-            self.sources[device.source_id] = device
-
+        source_id = config_entry.data[CONF_SOURCE_ID]
+        assert source_id not in self.sources
```
We only use assert in production code for type annotation to help mypy. I don't understand why we use it here.

chishm commented:
> We only use assert in production code for type annotation to help mypy.

Why is that? Is there documentation about this somewhere? 

> I don't understand why we use it here.

This is a case I don't ever expect to happen, but if it does it will be much harder to debug later (without the assertion). I'm asserting my conceptions about the code to both the interpreter and other developers.

MartinHjelmare commented:
Python assertion statements can be turned off so we can't rely on them for general runtime checks running on user hosts.

If this is only meant as a check during development, please add a comment explaining it.

chishm committed:
Comment for debug assert

chishm commented:
> Python assertion statements can be turned off so we can't rely on them for general runtime checks running on user hosts.
> 
> If this is only meant as a check during development, please add a code comment explaining it.

Thanks for the clarification. I've submitted a [PR to the developer docs](https://github.com/home-assistant/developers.home-assistant/pull/1255) to help spread the word.

MartinHjelmare reviewed (approved):
Thanks!
